src/main/java/isaac/bastion/storage/BastionGroupStorage.java

```diff
@@ -0,0 +1,295 @@
+		if(bastionGroups.size() > 0) {
+			bastionGroup = bastionGroups.get(0);
+		} else {
+			this.groups.put(group.getGroupId(), bastionGroup = new BastionGroup(group.getGroupId()));
+		}
+
+		bastionGroup.addAllowedGroup(allowedGroup.getGroupId());
+		addChanged(bastionGroup.getGroupId(), allowedGroup.getGroupId(), true);
+
+		return true;
+	}
+
+	/**
+	 * Deletes a bastion group
+	 * @param group The bastion's group
+	 * @param allowedGroup The allowed group name to delete
+	 * @return True is group was deleted, False - if group doesn't exist
+	 */
+	public boolean deleteAllowedGroup(Group group, Group allowedGroup) {
+		boolean isDeleted = false;
+		BastionGroupAndAllowed bastionGroupAndAllowed;
+
+		while((bastionGroupAndAllowed = getBastionGroupByAllowed(group, allowedGroup)) != null) {
+			if (!bastionGroupAndAllowed.bastionGroup.removeAllowedGroup(bastionGroupAndAllowed.allowedGroupId)) {
+				this.groups.remove(bastionGroupAndAllowed.bastionGroup.getGroupId());
+			}
+
+			addChanged(bastionGroupAndAllowed.bastionGroup.getGroupId(), bastionGroupAndAllowed.allowedGroupId, false);
+
+			isDeleted = true;
+		}
+
+		return isDeleted;
+	}
+
+	public boolean isAllowedGroup(Group group, Group allowedGroup) {
+		return getBastionGroupByAllowed(group, allowedGroup) != null;
+	}
+
+	public List<BastionGroup> getBastionGroups(Group group) {
+		List<BastionGroup> bastionGroups = new ArrayList<BastionGroup>();
+
+		for(int groupId : group.getGroupIds()) {
+			BastionGroup bastionGroup = this.groups.get(groupId);
+
+			if(bastionGroup != null) {
+				bastionGroups.add(bastionGroup);
+			}
+		}
+
+		return bastionGroups;
+	}
+
+
+	/**
+	 * Loads all bastion groups from the database
+	 */
+	public void loadGroups() {
+		try (
+			Connection conn = this.db.getConnection();
+			PreparedStatement ps = conn.prepareStatement(selectAllGroups)
+		) {
+			int validCount = 0;
+			int invalidCount = 0;
+			ResultSet result = ps.executeQuery();
+
+			while(result.next()) {
+				int groupId = result.getInt("bastion_group_id");
+				int allowedGroupId = result.getInt("allowed_group_id");
+				Group group = GroupManager.getGroup(groupId);
+				Group allowedGroup = GroupManager.getGroup(allowedGroupId);
+
+				if(group == null || allowedGroup == null) {
+					addChanged(groupId, allowedGroupId, false);
+					invalidCount++;
+				} else {
+					BastionGroup bastionGroup = this.groups.get(groupId);
+
+					if(bastionGroup == null) {
+						this.groups.put(groupId, bastionGroup = new BastionGroup(groupId));
+					}
+
+					bastionGroup.addAllowedGroup(allowedGroupId);
+
+					validCount++;
+				}
+			}
+
+			log.info("Loaded " + this.groups.size() + " bastion groups with " + validCount + " allowed groups.");
+
+			if(invalidCount > 0) {
+				log.info(" Marked to delete " + invalidCount + " records for non-existent groups.");
+			}
+		} catch (SQLException e) {
+			e.printStackTrace();
+			log.log(Level.SEVERE, "Error loading bastion groups from database, shutting down", e);
```

Comment:
For something as severe as shutting down the server? I'm not sure that's indicated here, but if it is, please, make the message a bit more ... dramatic. 

Red text perhaps? dramatic "=====" bars to separate top and bottom of report?